File: sdk/go/feature_flags.go
package alpgo

import (
	"fmt"
	"strings"
	"time"
)

// FeatureFlagEngine — v74.0.0 Feature Flag Engine
// Dynamic feature flags for agent workflows: gradual rollouts, percentage-based targeting, kill switches.

type FlagStatus string

const (
	FlagEnabled    FlagStatus = "ENABLED"
	FlagDisabled   FlagStatus = "DISABLED"
	FlagRollout    FlagStatus = "ROLLOUT"
	FlagExperiment FlagStatus = "EXPERIMENT"
)

type FlagVariant struct {
	VariantID string                 `json:"variant_id"`
	Name      string                 `json:"name"`
	Weight    int                    `json:"weight"`
	Payload   map[string]interface{} `json:"payload,omitempty"`
}

type FeatureFlag struct {
	FlagID             string        `json:"flag_id"`
	Name               string        `json:"name"`
	Description        string        `json:"description"`
	Status             FlagStatus    `json:"status"`
	RolloutPercentage  int           `json:"rollout_percentage"`
	TargetAgents       []string      `json:"target_agents"`
	TargetEnvironments []string      `json:"target_environments"`
	Variants           []FlagVariant `json:"variants"`
	KillSwitch         bool          `json:"kill_switch"`
	CreatedAt          string        `json:"created_at"`
	UpdatedAt          string        `json:"updated_at"`
}

type FlagEvaluation struct {
	FlagID      string       `json:"flag_id"`
	AgentID     string       `json:"agent_id"`
	Environment string       `json:"environment"`
	Enabled     bool         `json:"enabled"`
	Variant     *FlagVariant `json:"variant,omitempty"`
	Reason      string       `json:"reason"`
}

type FeatureFlagEngine struct {
	flags map[string]*FeatureFlag
}

func NewFeatureFlagEngine() *FeatureFlagEngine {
	return &FeatureFlagEngine{
		flags: make(map[string]*FeatureFlag),
	}
}

func (ffe *FeatureFlagEngine) CreateFlag(name, description string, status FlagStatus, rollout int) *FeatureFlag {
	id := fmt.Sprintf("flag-%s-%d", strings.ToLower(strings.ReplaceAll(name, " ", "-")), time.Now().UnixNano())
	now := time.Now().UTC().Format(time.RFC3339)
	flag := &FeatureFlag{
		FlagID:            id,
		Name:              name,
		Description:       description,
		Status:            status,
		RolloutPercentage: rollout,
		TargetAgents:      []string{},
		TargetEnvironments: []string{},
		Variants:          []FlagVariant{},
		KillSwitch:        false,
		CreatedAt:         now,
		UpdatedAt:         now,
	}
	ffe.flags[id] = flag
	return flag
}

func (ffe *FeatureFlagEngine) Evaluate(flagID, agentID, environment string) *FlagEvaluation {
	flag, ok := ffe.flags[flagID]
	if !ok {
		return &FlagEvaluation{FlagID: flagID, AgentID: agentID, Environment: environment, Enabled: false, Reason: "FLAG_NOT_FOUND"}
	}
	if flag.KillSwitch {
		return &FlagEvaluation{FlagID: flagID, AgentID: agentID, Environment: environment, Enabled: false, Reason: "KILL_SWITCH"}
	}
	if flag.Status == FlagDisabled {
		return &FlagEvaluation{FlagID: flagID, AgentID: agentID, Environment: environment, Enabled: false, Reason: "FLAG_DISABLED"}
	}
	return &FlagEvaluation{FlagID: flagID, AgentID: agentID, Environment: environment, Enabled: true, Reason: "FLAG_ENABLED"}
}

func (ffe *FeatureFlagEngine) GetFlags() []*FeatureFlag {
	res := make([]*FeatureFlag, 0, len(ffe.flags))
	for _, f := range ffe.flags {
		res = append(res, f)
	}
	return res
}

File: sdk/rust/src/feature_flags.rs
use std::collections::HashMap;
use chrono::Utc;

/// v74.0.0 Feature Flag Engine
/// Dynamic feature flags for agent workflows: gradual rollouts, percentage-based targeting, kill switches.

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
pub enum FlagStatus {
    Enabled,
    Disabled,
    Rollout,
    Experiment,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct FlagVariant {
    pub variant_id: String,
    pub name: String,
    pub weight: u32,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct FeatureFlag {
    pub flag_id: String,
    pub name: String,
    pub description: String,
    pub status: FlagStatus,
    pub rollout_percentage: u32,
    pub kill_switch: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct FlagEvaluation {
    pub flag_id: String,
    pub agent_id: String,
    pub environment: String,
    pub enabled: bool,
    pub reason: String,
}

pub struct FeatureFlagEngine {
    flags: HashMap<String, FeatureFlag>,
}

impl FeatureFlagEngine {
    pub fn new() -> Self {
        Self { flags: HashMap::new() }
    }

    pub fn create_flag(&mut self, name: &str, description: &str, status: FlagStatus, rollout: u32) -> &FeatureFlag {
        let id = format!("flag-{}", Utc::now().timestamp_nanos_opt().unwrap_or(0));
        let now = Utc::now().to_rfc3339();
        let flag = FeatureFlag {
            flag_id: id.clone(),
            name: name.to_string(),
            description: description.to_string(),
            status,
            rollout_percentage: rollout,
            kill_switch: false,
            created_at: now.clone(),
            updated_at: now,
        };
        self.flags.insert(id.clone(), flag);
        self.flags.get(&id).unwrap()
    }

    pub fn evaluate(&self, flag_id: &str, agent_id: &str, environment: &str) -> FlagEvaluation {
        match self.flags.get(flag_id) {
            None => FlagEvaluation {
                flag_id: flag_id.to_string(),
                agent_id: agent_id.to_string(),
                environment: environment.to_string(),
                enabled: false,
                reason: "FLAG_NOT_FOUND".to_string(),
            },
            Some(flag) => {
                if flag.kill_switch {
                    return FlagEvaluation {
                        flag_id: flag_id.to_string(),
                        agent_id: agent_id.to_string(),
                        environment: environment.to_string(),
                        enabled: false,
                        reason: "KILL_SWITCH".to_string(),
                    };
                }
                if flag.status == FlagStatus::Disabled {
                    return FlagEvaluation {
                        flag_id: flag_id.to_string(),
                        agent_id: agent_id.to_string(),
                        environment: environment.to_string(),
                        enabled: false,
                        reason: "FLAG_DISABLED".to_string(),
                    };
                }
                FlagEvaluation {
                    flag_id: flag_id.to_string(),
                    agent_id: agent_id.to_string(),
                    environment: environment.to_string(),
                    enabled: true,
                    reason: "FLAG_ENABLED".to_string(),
                }
            }
        }
    }

    pub fn get_flags(&self) -> Vec<&FeatureFlag> {
        self.flags.values().collect()
    }
}
